### PR TITLE
AMIS Integration

### LINKED ISSUE #32: Create script for running the model through AMIS

### PR DESCRIPTION
**Pull request by thk123**
Implementing the wrapper layer between the model and AMIS. This is inside a folder called amis_integration inside sch_simulation. It contains a README explaining how to use

**The Model Wrapper**

Duplicates a lot of the logic of Example_Run.py, but exposes all the customisable points so that the R script dictates the run. 

The entry point is `run_model_with_parameters` which is expected to be called from R with appropriate parameters. It runs on a single thread - this might be something that we want to change. 

**The R code**

There is a script `sth_fitting.R` which contains the bespoke parameters for a specific fit (one could imagine having another script called `sch_fitting.R` with different parameters. It uses `amis_integration.R` which handles the python/R interface, as well as making sure it uses the right Python virtual env. 

At the moment it contains totally dummy prior, prevalence maps etc - this will be done by Evandro.

Fixes NTD-Modelling-Consortium/endgame-project#3 

To do:

 - [x] Get the virtual environment programmatically in the R script
 - [x] Allow the R script to drive what years we are trying to fit to
 - [x] Fix typo in name of the R or Python script (whichever is wrong!)
 - [x] Reference specific AMIS version
 - [x] Add a trivial R test to check the utility 
 - [x] Fix python version (3.8 does not provide the `@cache` functionality
 - [x] Fix the import
 - [x] Test on 3.10 rather than 3.12
 - [x] Ensure cache is working

### CODE REVIEW

File: MANIFEST.in [hunk 1]
```diff
@@ -2,3 +2,4 @@ include sch_simulation/data/*.txt
 include sch_simulation/data/*.xlsx
 include sch_simulation/data/*/*.txt
 include sch_simulation/data/*/*.xlsx
+include sch_simulation/data/*/*.pickle
```
Comment (thk123) [inline MANIFEST.in:5]: This adds the random pickle file in the data directory to the installed package - which makes it easier to access the file using the pkg_resource library. However, should think about whether precomputed simulated data really does belong in the repo at all. Alsom the `pkg_resource` library is deprecated, so maybe should look at a different way to manage this in any case.

**Comment by thk123** [comment]
@mattg3004 I've put this back as a draft as I've realised there are a couple of tidy up things I should do that were caught by Evandro trying to use this, so I'll do that and then ping you when it is finally ready for review.

**Comment by thk123** [comment]
Raised NTD-Modelling-Consortium/endgame-project#14 for the multi point fitting

File: pickle_example.py [hunk 2]
```diff
@@ -72,7 +72,7 @@
 res = Parallel(n_jobs=num_cores)(
          delayed(multiple_simulations)(params, pickleData, simparams, indices, i) for i in range(numSims))
 results=[item[0] for item in res]
-simData=[item[1] for item in res]
+simData:list[SDEquilibrium]=[item[1] for item in res]
```
Comment (thk123) [inline pickle_example.py:75]: This change does not belong in this PR.

File: MANIFEST.in (hunk 1 above)
Comment (thk123) [reply MANIFEST.in:5]: I don't think this change is required and so should be dropped for now

File: sch_simulation/amis_integration/.gitignore [hunk 3]
```diff
@@ -0,0 +1,3 @@
+renv/**
+Man_MDA_vacc.txt
+.Rprofile
```
Comment (thk123) [inline sch_simulation/amis_integration/.gitignore:3]: Missing new line

File: sch_simulation/amis_integration/README.md [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+Make sure you've set up the Python virtual environment for the
+sch_simulation package (see main [README](../../README.md)).
+
+1. Launch a R session `R`
+2. Create a R virtual environment `renv::init()`
+3. Install dependencies `renv::restore()`
+
+## Running
+
```
Comment (thk123) [inline sch_simulation/amis_integration/README.md:24]: Put about working directory

File: sch_simulation/amis_integration/amis_integration.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    PrevalenceEstimate = getPrevalenceWholePop(
+        output, params, numReps, params.Unfertilized, fixed_parameters.survey_type, 1
+    )
+    return PrevalenceEstimate
+
+
+def extract_relevant_results(results: pandas.DataFrame) -> float:
+    num_years = len(results)
+    # TODO: for now we assume we are interested in fitting the last year
+    return results["draw_1"][num_years - 1]
```
Comment (thk123) [inline sch_simulation/amis_integration/amis_integration.py:99]: ❓ @mattg3004 Is this right? Here I think "draw_1" means we're just running one simulation

File: sch_simulation/amis_integration/amis_integration.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+
+def run_model_with_parameters(seeds, parameters, fixed_parameters: FixedParameters):
```
Comment (thk123) [inline sch_simulation/amis_integration/amis_integration.py:102]: ❓ We don't use the seed - is that correct?

File: sch_simulation/amis_integration/reticulate_util.R [hunk 7]
```diff
@@ -0,0 +1,16 @@
+get_venv <- function() {
```
Comment (thk123) [inline sch_simulation/amis_integration/reticulate_util.R:1]: Here we are assuming pipenv but this isn't standard and shouldn't be enforced but looks like reticulate does try and use the activating virtual environement anyway so can perhaps just delete this.

File: sch_simulation/amis_integration/tests/testthat.R [hunk 8]
```diff
@@ -0,0 +1,12 @@
+# This file is part of the standard setup for testthat.
+# It is recommended that you do not modify it.
+#
+# Where should you do additional test configuration?
+# Learn more about the roles of various files in:
+# * https://r-pkgs.org/testing-design.html#sec-tests-files-overview
+# * https://testthat.r-lib.org/articles/special-files.html
+
+print("Running tests")
```
Comment (thk123) [inline sch_simulation/amis_integration/tests/testthat.R:9]: Delete debug print

File: sch_simulation/amis_integration/tests/testthat.R [hunk 9]
```diff
@@ -0,0 +1,12 @@
+# This file is part of the standard setup for testthat.
```
Comment (thk123) [inline sch_simulation/amis_integration/tests/testthat.R:1]: Delete this comment - I even have modified it.

File: sch_simulation/amis_integration/amis_integration.py [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+
+    # run a single realization
+    results, SD = sch_simulation.helsim_RUN_KK.doRealizationSurveyCoveragePickle(
+        params,
+        fixed_parameters.survey_type,
+        simData,
+        fixed_parameters.min_multiplier,
+    )
+
+    results = [results]
```
Comment (thk123) [inline sch_simulation/amis_integration/amis_integration.py:84]: ❓ Matt - I took this but why do we need to wrap results in a list?

Comment (mattg3004) [reply sch_simulation/amis_integration/amis_integration.py:84]: the function extractHostData which is called next expects a list of lists. Running one rep of the doRealizationSurveyCoveragePickle as given in the example returns a have a list of sch_simulation.helsim_FUNC_KK.helsim_structures.Result. 
We can get around this by wrapping in a list here. Better would be to edit the extractHostData function to test whether we would need to wrap in a list, or have a separate process in the function if it isn't a list of lists

File: sch_simulation/amis_integration/amis_integration.py (hunk 5 above)
Comment (mattg3004) [reply sch_simulation/amis_integration/amis_integration.py:99]: Yes, this is kind of true. Here draw_1 is just a dummy name for the results of the simulation. When we run the endgame simulations, we produce a file at the end with draw_0, draw_1,..., draw_199, as we are running 200 simulations. However, all of these different columns come from individual runs which are output with a column called draw_1 and then these are all combined and renamed draw_0-draw_199.

In this example, we could instead of running a single simulation, replace this with:

numSims = 2
res=Parallel(n_jobs=num_cores)(
delayed(doRealizationSurveyCoveragePickle)(params, surveyType, simData, mult) for i in range(numSims))
results=[item[0] for item in res]
and then run extractHostData and getPrevalenceWholePop and this would still return an output with just a draw_1 column, as the result would be the mean of the 2 runs

File: sch_simulation/amis_integration/amis_integration.py (hunk 6 above)
Comment (mattg3004) [reply sch_simulation/amis_integration/amis_integration.py:102]: currently no. It would be useful to have a seed for the fitting though I think, at least the option to have one, so we should add this

Comment (thk123) [reply sch_simulation/amis_integration/amis_integration.py:102]: I have raised ntd-modelling-consortium/endgame-project#18 for doing this

File: sch_simulation/amis_integration/amis_integration.py (hunk 10 above)
Comment (thk123) [reply sch_simulation/amis_integration/amis_integration.py:84]: Ah that makes sense, thank you. I think it is fine for the function to expect a specific format - what I'll do to make it clearer why this is happening is wrap it when we actually call the `extractHostData`:

```python
extractHostData([results]) 
```

File: sch_simulation/amis_integration/amis_integration.py (hunk 5 above)
Comment (thk123) [reply sch_simulation/amis_integration/amis_integration.py:99]: OK thanks - I'll pull out this magic string at least from this function - ideally it would be just one place to avoid discrepancies. Is there a reason to use `draw_1` rather than say `prevalence` if it is always just one draw?

**Comment by thk123** [comment]
@mattg3004 I am happy for this to be merged, if you are happy with (ie understand it, can't see any problems), then could you "approve" it, by going to the files changed tab and submit an approving review: 
image
Thanks 💚

**Review by mattg3004 — APPROVED** [review]
I don't know what is going on with the amis functions, should I be looking into these in order to properly review this, or am I just on the python code?
Otherwise all looks good to me.

**Comment by thk123** [comment]
@mattg3004 Thanks for reviewing - I've asked Raiha to review the R side of things.

**Review by RaihaTuiTaura — COMMENTED** [review]
R side of the AMIS integration looks good to me.

File: sch_simulation/amis_integration/amis_integration.R [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+build_transmission_model <- function(prevalence_map, fixed_parameters) {
+  sch_simulation <- get_amis_integration_package()
+  transmission_model <- function(seeds, params, n_tims = length(prevalence_map)) {
+    output <- sch_simulation$run_model_with_parameters(
+      seeds, params, fixed_parameters
+    )
+    print("Output:")
+    print(output)
```
[the remark this answers is not in the document]
Comment (thk123) [reply sch_simulation/amis_integration/amis_integration.R]: Good point - I've removed this and another print statement. A progress bar does sound nice but I don't know how to do it quickly - happy to raise an issue if the lack of progress feedback impacts being able to effectively work on the fitting - I'm sure it won't be too complicated.
